### ISSUE TITLE
Recent ps3netsrv Linux releases don't work at all

### ISSUE THREAD

**Issue by kam821** [aldostools/webMAN-MOD#366]
@aldostools 
I took a break from using PS3, so I didn't notice the problem immediately.
Yesterday, I updated ps3netsrv to version 20200529.
PS3 does load remote file list correctly after XML Refresh, but nothing happens when I try to load the game, i.e. the disc image is not mounted / is not visible in the game list.
ps3netsrv output displays path opening attempt, server doen't crash.
It doesn't matter it's precompiled binary or static (Make.sh) / dynamic (meson) self-compiled version.
If I remember correctly, error 404 /dev/bdvd not found was displayed after trying to mount any remote game in webMAN webpanel.
I tested every recent release and and the latest working version is 20200520A.
System: Debian 10 / gcc 8.3.0-6
Model: CECH-2504B
Firmware: 4.84.2 REX (DEX kernel) / Cobra 8.2
Webman-MOD 1.47.28 Cobra-PS3MAPI ( release: https://github.com/aldostools/webMAN-MOD/commit/f7c9dfb86d13c0915a2eb86fdc69095e35712268)

**Comment by kam821** [comment]
@aldostools 
btw: There is only one error left in the latest version compilation.
```
../src/VIsoFile.cpp:1665:18: warning: comparison of integer expressions of different signedness: 'int64_t' {aka 'long int'} and 'uint64_t' {aka 'long unsigned int'} [-sign-compare ]
        if (this_r! = to_read)
            ^ ~~~~~~~ ~~~~~~~~~
```
Personally, I'm a total noob, but I'm guessing, that you could fix it by converting this_r to unsigned in comparing function, because just before it, there is return if this_r is negative number.
Am I right or am I wrong?
```
if (this_r < 0)
{
        fprintf(stderr, "VISO: read_file failed on %s\n", fileList->path);
        return r;
}
if ((uint64_t)this_r != to_read)
{
        fprintf(stderr, "VISO: read on file %s returned less data than expected (file modified?)\n", fileList->path);
        return r;
}

```

**Comment by aldostools** [comment]
About the fix for error, your fix should work, although it's suggested to change the unsigned to signed always that the possible values don't exceed the limit of the data range. 

Here is an hybrid of 20200529 with 20200520A (basically VisoFile.cpp and VisoFile.h were replaced with the original version)
[ps3netsrv_20200604.zip](https://github.com/aldostools/webMAN-MOD/files/4730398/ps3netsrv_20200604.zip)

**Comment by kam821** [comment]
Weird.
This version behaves identically to what I wrote in the main post, so it also doesn't work.
I forgot to add earlier that the first attempt to load the game causes server displays an attempt to open the iso file, but each subsequent attempt doesn't make any output in ps3netsrv, only console reboot helps.

**Comment by aldostools** [comment]
Meanwhile use this:
[ps3netsrv_20200520A.zip](https://github.com/aldostools/webMAN-MOD/files/4730623/ps3netsrv_20200520A.zip)

**Comment by jhonathanc** [comment]
The ps3netsrv I compiled has the same weird behavior. I'm using the v20200607.

**Comment by aldostools** [comment]
Thank you for your feedback. I've tested the project on CYGWIN and it works ok.

What distro of linux are you using? 
What version of GCC and G++ are you using? 
There is an issue #353 reported building the project with GCC 10. If so, try with GCC 9.3

Does ps3netsrv_20200520A works or have the same problem as 20200607?

Please be more specific what does not work:
- Not start?
- Start but don't connect?
- Connect but don't list directories?
- Connect but don't mount ISO games?
- Connect but don't mount folder games?
- Mount ISO game. but game don't start?
- Mount folder game. but game don't start?
- Game start, but ps3netsrv crashes?
- dev_bdvd show corrupted directories?

More detail help to know where to look.

**Comment by jhonathanc** [comment]
1. I'm using OpenWrt (ath79 target);
2. gcc version: 7.4.0-5 (did't found g++ version in OpenWrt toolchain);
3. PS3 connects to ps3netsrv and it's able to list games and show directory content;
3. The games are listed under webMan PS3 Games, but when I try to mount it, nothing occurs. (it returns to games category, but nothing happens) - same as reported by @kam821.
4. I got error 400 when I try to open the ISO navigating from webMan, on net0.
5. v20200520A works.

So, it connect but don't mount ISO games.

I will try to debug it later, I think I can reproduce the same error on Ubuntu.

Edit: The same behavior occurs on Ubuntu 20.04, GCC/G++ version 9.3.0. So, maybe it will also happen on all other (*)nix systems.

I will try to fix and submit a PR.

**Comment by aldostools** [comment]
@jhonathanc Thank you for your help I really appreaciate it.

Unfortunately my VM with linux (mint) cannot connect to my PS3, so I cannot test it myself.  

Here are 4 test versions with latest code but using files from 20200520A.
[ps3netsrv_test.zip](https://github.com/aldostools/webMAN-MOD/files/4755072/ps3netsrv_test.zip)

The major change in these 4 test versions is that the full lib of POLARSSL is not longer required.
Only the AES & padlock modules are used.

If you have a chance, please test them and report back which one works.
It should help to identify which module is causing the regression.

**Comment by jhonathanc** [comment]
Only ps3netsrv_20200607_viso don't work.

**Comment by kam821** [comment]
@aldostools 
Thanks for your hard work.
Finally, 3/4 test version work.

I'm attaching compilation output for every release.
- ps3netsrv_20200607 [original release] (doesn't work)
https://pastebin.com/iHMWYJXx
- ps3netsrv_20200607_file (works)
https://pastebin.com/x11umKFf 
- ps3netsrv_20200607_viso (failed, same behaviour as original, recent releases)
https://pastebin.com/pmqfCtCf
- ps3netsrv_20200607_viso_file (works)
https://pastebin.com/fwQUYtSg
- ps3netsrv_20200607_viso_file_main (works)
https://pastebin.com/AckJCB6x

I noticed, that main difference between working and non-working versions is that working releases displays warning during compilation File.cpp.
```
g++  -Wall -I./include -D_LARGEFILE64_SOURCE -D_FILE_OFFSET_BITS=64 -DPOLARSSL -Doff64_t=off_t  -c -o src/File.o src/File.cpp
src/File.cpp: In member function ‘virtual int File::open(const char*, int)’:
src/File.cpp:116:37: warning: passing argument 1 to restrict-qualified parameter aliases with argument 3 [-Wrestrict]
     filepath[flen + 4] = 0; sprintf(filepath, "%s.%i", filepath, i);
                                     ^~~~~~~~           ~~~~~~~~
```

About your previous questions.
System: Linux Debian 10, gcc/g++ 8.3.0, LXC unprivileged container, but temporary I tried to run non-working releases on physical Proxmox (Debian 10) and there wasn't difference.
Even "non-working" releases correctly sent files/games list to the PS3 XML during Refresh
But after you choose game in Webman Games, menu returned to main XMB/Game Category (as always), but nothing more happen.
No optical drive icon (ISO mounted as BD) as it should be.

But it looks, that this testing releases is turn in the right direction.
Thanks again :)

PS: I'm attaching complete diff between old, working File.cpp and new, not so much working.
https://pastebin.com/MnNA3f3n

**Comment by aldostools** [comment]
Thank you guy for the detailed test.

From the tests results it seems the error was caused by the recent changes to `File.cpp`

The error line 406: `if(offset == last_seek) return SUCCEEDED;`

This build is almost identical to 20200607 but without the "file seek" optimization on linux.
[ps3netsrv_20200609.zip](https://github.com/aldostools/webMAN-MOD/files/4755636/ps3netsrv_20200609.zip)

It uses as base this report: "ps3netsrv_20200607_file (works)" that only have File.cpp replaced.

**Comment by jhonathanc** [comment]
Yeah, this worked!

**Comment by aldostools** [comment]
> Yeah, this worked!

Thanks... I will commit 20200609 and release it. 

@kam821 You can close the issue if you think it's solved.

**Comment by nl6720** [comment]
[20200609](https://github.com/aldostools/webMAN-MOD/releases/download/1.47.28/ps3netsrv_20200609.zip) doesn't build:
<details><summary>log</summary>

```
The Meson build system
Version: 0.54.2
Source dir: /build/ps3netsrv/src/ps3netsrv_20200609
Build dir: /build/ps3netsrv/src/ps3netsrv_20200609/build
Build type: native build
Project name: ps3netsrv
Project version: 0.1
Using 'CFLAGS' from environment with value: '-march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fcommon'
Using 'LDFLAGS' from environment with value: '-Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now'
Using 'CPPFLAGS' from environment with value: '-D_FORTIFY_SOURCE=2'
Using 'CXXFLAGS' from environment with value: '-march=x86-64 -mtune=generic -O2 -pipe -fno-plt'
Using 'LDFLAGS' from environment with value: '-Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now'
Using 'CPPFLAGS' from environment with value: '-D_FORTIFY_SOURCE=2'
Using 'CFLAGS' from environment with value: '-march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fcommon'
Using 'LDFLAGS' from environment with value: '-Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now'
Using 'CPPFLAGS' from environment with value: '-D_FORTIFY_SOURCE=2'
C compiler for the host machine: cc (gcc 10.1.0 "cc (GCC) 10.1.0")
C linker for the host machine: cc ld.bfd 2.34.0
Using 'CXXFLAGS' from environment with value: '-march=x86-64 -mtune=generic -O2 -pipe -fno-plt'
Using 'LDFLAGS' from environment with value: '-Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now'
Using 'CPPFLAGS' from environment with value: '-D_FORTIFY_SOURCE=2'
C++ compiler for the host machine: c++ (gcc 10.1.0 "c++ (GCC) 10.1.0")
C++ linker for the host machine: c++ ld.bfd 2.34.0
Host machine cpu family: x86_64
Host machine cpu: x86_64
Run-time dependency threads found: YES
Library mbedtls found: YES
Library mbedcrypto found: YES
Build targets in project: 1

Found ninja-1.10.0 at /usr/bin/ninja
ninja: Entering directory `build'
[1/5] Compiling C object 'ps3netsrv@exe/src_compat.c.o'
[2/5] Compiling C++ object 'ps3netsrv@exe/src_File.cpp.o'
FAILED: ps3netsrv@exe/src_File.cpp.o 
c++ -Ips3netsrv@exe -I. -I.. -I../include -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -std=c++11 -march=x86-64 -mtune=generic -O2 -fno-plt -D_FORTIFY_SOURCE=2 -pthread -MD -MQ 'ps3netsrv@exe/src_File.cpp.o' -MF 'ps3netsrv@exe/src_File.cpp.o.d' -o 'ps3netsrv@exe/src_File.cpp.o' -c ../src/File.cpp
In file included from ../src/File.cpp:1:
../include/File.h:112:20: error: ‘mbedtls_aes_context’ has not been declared
  112 |  void decrypt_data(mbedtls_aes_context &aes, unsigned char *data, int sector_count, unsigned int start_lba);
      |                    ^~~~~~~~~~~~~~~~~~~
../include/File.h:121:2: error: ‘mbedtls_aes_context’ does not name a type
  121 |  mbedtls_aes_context aes_dec_;
      |  ^~~~~~~~~~~~~~~~~~~
../src/File.cpp: In member function ‘virtual int File::open(const char*, int)’:
../src/File.cpp:180:32: error: ‘aes_dec_’ was not declared in this scope
  180 |    if (mbedtls_aes_setkey_dec(&aes_dec_, key, 128) == SUCCEEDED)
      |                                ^~~~~~~~
../src/File.cpp:180:8: error: ‘mbedtls_aes_setkey_dec’ was not declared in this scope; did you mean ‘aes_setkey_dec’?
  180 |    if (mbedtls_aes_setkey_dec(&aes_dec_, key, 128) == SUCCEEDED)
      |        ^~~~~~~~~~~~~~~~~~~~~~
      |        aes_setkey_dec
../src/File.cpp:248:4: error: ‘mbedtls_aes_context’ was not declared in this scope
  248 |    mbedtls_aes_context aes_d1;
      |    ^~~~~~~~~~~~~~~~~~~
../src/File.cpp:249:32: error: ‘aes_d1’ was not declared in this scope; did you mean ‘key_d1’?
  249 |    if (mbedtls_aes_setkey_enc(&aes_d1, key_d1, 128) == SUCCEEDED)
      |                                ^~~~~~
      |                                key_d1
../src/File.cpp:249:8: error: ‘mbedtls_aes_setkey_enc’ was not declared in this scope; did you mean ‘aes_setkey_enc’?
  249 |    if (mbedtls_aes_setkey_enc(&aes_d1, key_d1, 128) == SUCCEEDED)
      |        ^~~~~~~~~~~~~~~~~~~~~~
      |        aes_setkey_enc
../src/File.cpp:250:40: error: ‘MBEDTLS_AES_ENCRYPT’ was not declared in this scope
  250 |     if (mbedtls_aes_crypt_cbc(&aes_d1, MBEDTLS_AES_ENCRYPT, 16, &iv_d1[0], key, key) == SUCCEEDED)
      |                                        ^~~~~~~~~~~~~~~~~~~
../src/File.cpp:250:9: error: ‘mbedtls_aes_crypt_cbc’ was not declared in this scope
  250 |     if (mbedtls_aes_crypt_cbc(&aes_d1, MBEDTLS_AES_ENCRYPT, 16, &iv_d1[0], key, key) == SUCCEEDED)
      |         ^~~~~~~~~~~~~~~~~~~~~
../src/File.cpp:251:34: error: ‘aes_dec_’ was not declared in this scope
  251 |      if (mbedtls_aes_setkey_dec(&aes_dec_, key, 128) == SUCCEEDED)
      |                                  ^~~~~~~~
../src/File.cpp:251:10: error: ‘mbedtls_aes_setkey_dec’ was not declared in this scope; did you mean ‘aes_setkey_dec’?
  251 |      if (mbedtls_aes_setkey_dec(&aes_dec_, key, 128) == SUCCEEDED)
      |          ^~~~~~~~~~~~~~~~~~~~~~
      |          aes_setkey_dec
../src/File.cpp: In member function ‘virtual ssize_t File::read(void*, size_t)’:
../src/File.cpp:349:18: error: ‘aes_dec_’ was not declared in this scope
  349 |     decrypt_data(aes_dec_, reinterpret_cast<unsigned char *>(buf), ret / kSectorSize, read_position / kSectorSize);
      |                  ^~~~~~~~
../src/File.cpp: At global scope:
../src/File.cpp:497:6: error: variable or field ‘decrypt_data’ declared void
  497 | void File::decrypt_data(mbedtls_aes_context &aes, unsigned char *data, int sector_count, unsigned int start_lba)
      |      ^~~~
../src/File.cpp:497:25: error: ‘mbedtls_aes_context’ was not declared in this scope
  497 | void File::decrypt_data(mbedtls_aes_context &aes, unsigned char *data, int sector_count, unsigned int start_lba)
      |                         ^~~~~~~~~~~~~~~~~~~
../src/File.cpp:497:46: error: ‘aes’ was not declared in this scope; did you mean ‘abs’?
  497 | void File::decrypt_data(mbedtls_aes_context &aes, unsigned char *data, int sector_count, unsigned int start_lba)
      |                                              ^~~
      |                                              abs
../src/File.cpp:497:51: error: expected primary-expression before ‘unsigned’
  497 | void File::decrypt_data(mbedtls_aes_context &aes, unsigned char *data, int sector_count, unsigned int start_lba)
      |                                                   ^~~~~~~~
../src/File.cpp:497:72: error: expected primary-expression before ‘int’
  497 | void File::decrypt_data(mbedtls_aes_context &aes, unsigned char *data, int sector_count, unsigned int start_lba)
      |                                                                        ^~~
../src/File.cpp:497:90: error: expected primary-expression before ‘unsigned’
  497 | void File::decrypt_data(mbedtls_aes_context &aes, unsigned char *data, int sector_count, unsigned int start_lba)
      |                                                                                          ^~~~~~~~
[3/5] Compiling C++ object 'ps3netsrv@exe/src_VIsoFile.cpp.o'
FAILED: ps3netsrv@exe/src_VIsoFile.cpp.o
c++ -Ips3netsrv@exe -I. -I.. -I../include -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -std=c++11 -march=x86-64 -mtune=generic -O2 -fno-plt -D_FORTIFY_SOURCE=2 -pthread -MD -MQ 'ps3netsrv@exe/src_VIsoFile.cpp.o' -MF 'ps3netsrv@exe/src_VIsoFile.cpp.o.d' -o 'ps3netsrv@exe/src_VIsoFile.cpp.o' -c ../src/VIsoFile.cpp
In file included from ../include/VIsoFile.h:7,
                 from ../src/VIsoFile.cpp:8:
../include/File.h:112:20: error: ‘mbedtls_aes_context’ has not been declared
  112 |  void decrypt_data(mbedtls_aes_context &aes, unsigned char *data, int sector_count, unsigned int start_lba);
      |                    ^~~~~~~~~~~~~~~~~~~
../include/File.h:121:2: error: ‘mbedtls_aes_context’ does not name a type
  121 |  mbedtls_aes_context aes_dec_;
      |  ^~~~~~~~~~~~~~~~~~~
[4/5] Compiling C++ object 'ps3netsrv@exe/src_main.cpp.o'
FAILED: ps3netsrv@exe/src_main.cpp.o
c++ -Ips3netsrv@exe -I. -I.. -I../include -fdiagnostics-color=always -pipe -D_FILE_OFFSET_BITS=64 -std=c++11 -march=x86-64 -mtune=generic -O2 -fno-plt -D_FORTIFY_SOURCE=2 -pthread -MD -MQ 'ps3netsrv@exe/src_main.cpp.o' -MF 'ps3netsrv@exe/src_main.cpp.o.d' -o 'ps3netsrv@exe/src_main.cpp.o' -c ../src/main.cpp
In file included from ../src/main.cpp:23:
../include/File.h:112:20: error: ‘mbedtls_aes_context’ has not been declared
  112 |  void decrypt_data(mbedtls_aes_context &aes, unsigned char *data, int sector_count, unsigned int start_lba);
      |                    ^~~~~~~~~~~~~~~~~~~
../include/File.h:121:2: error: ‘mbedtls_aes_context’ does not name a type
  121 |  mbedtls_aes_context aes_dec_;
      |  ^~~~~~~~~~~~~~~~~~~
ninja: build stopped: subcommand failed.
```

</details>

**Comment by aldostools** [comment]
@nl6720 The header for `<mbedtls/aes.h>` was left commented by mistake in `File.h`
(it was commented during the tests for removal of the dependency of POLARSSL)

Please try this update.: [ps3netsrv_20200610.zip](https://github.com/aldostools/webMAN-MOD/files/4757262/ps3netsrv_20200610.zip)

**Comment by nl6720** [comment]
20200610 works! Both folder and ISO games.

**Comment by kam821** [comment]
@aldostools  @jhonathanc  @nl6720 
I confirm that everything works fine, thank you so much ;)
mbedtls header error which prevented compilation using meson is also fixed in 20200610
both dynamic and static versions compile without warnings (feels good man)

I have some suggestions left, but not directly related to ps3netsrv.
- It would be cool if you could put all ps3netsrv files in directory named the same name as zip before compression (currently cygwin1.dll LICENSE.TXT MSCOMCTL.OCX MSVBVM50.DLL PS3_NET_Server.exe ps3netsrv.exe are extracted directly into working directory).
- fix line-breaking type in zip releases i.e. LF format instead of CRLF for Linux files (currently conversion is needed before executing, e.g. for Make.sh)
